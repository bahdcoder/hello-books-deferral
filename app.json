{
  "name": "Hello Books",
  "description": "A system that allows efficient, cost-effective management of Books for Libraries."
}
