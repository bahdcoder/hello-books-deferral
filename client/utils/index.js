export const getDateStr = str => {
    const end = str.indexOf('T');
    const rawDate = str.slice(0, end);
    const firstHyphen = rawDate.indexOf('-');
    const lastHyphen = rawDate.lastIndexOf('-');
    const year = rawDate.slice(0, firstHyphen);
    const month = rawDate.slice(firstHyphen + 1, lastHyphen);
    const day = rawDate.slice(lastHyphen + 1);
    const monthStrArr = [
        'Jan',
        'Feb',
        'Mar',
        'Apr',
        'May',
        'Jun',
        'Jul',
        'Aug',
        'Sep',
        'Oct',
        'Nov',
        'Dec'
    ];
    const monthStr = monthStrArr[Number(month) - 1];
    return `${monthStr} ${day} ${year}`;
};
