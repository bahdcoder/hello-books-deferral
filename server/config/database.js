export default () => ({
    development: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    test: {
        client: process.env.TEST_DATABASE_CLIENT,
        connection: process.env.TEST_DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    review: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.BACKUP_DATABASE,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    staging: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    production: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    }
});
