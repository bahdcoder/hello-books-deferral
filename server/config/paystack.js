export default () => ({
    secret: process.env.PAYSTACK_SECRET_KEY,
    public: process.env.PAYSTACK_PUBLIC_KEY
});
