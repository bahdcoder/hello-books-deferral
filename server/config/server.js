export default () => ({
  port: process.env.PORT || 5000,
  env: process.env.NODE_ENV || "development",
  url: process.env.APP_URL || "http://localhost:5000"
});
