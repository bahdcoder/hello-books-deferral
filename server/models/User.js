const { Model } = require("objection");
const bcrypt = require("bcrypt");

class User extends Model {
  static get tableName() {
    return "users";
  }
  async $beforeInsert(context) {
    await super.$beforeInsert(context);
    const salt = await bcrypt.genSalt(10);
    this.password = await bcrypt.hash(this.password, salt);
  }
}

module.exports = User;
