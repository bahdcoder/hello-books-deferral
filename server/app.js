import express from 'express';
import consola from 'consola';
import bodyParser from 'body-parser';
import swagger from 'swagger-ui-express';
import Knex from 'knex';
import { createServer } from 'http';
import { Model } from 'objection';
import swaggerDocument from '@/swagger.json';
import Routes from '@routes/v1';
import config from '@config';

import 'express-jsend';

export const app = express();

export const databaseConnection = Knex(config.database[config.server.env]);

Model.knex(databaseConnection);

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

Routes(app);

app.get('/', (req, res) => {
    res.status(200).jsend({ message: 'Hello Books Deferral' });
});
/**
 *  setup Swagger
 */
app.use(
    '/api-docs',
    swagger.serve,
    swagger.setup(swaggerDocument, { explorer: true })
);

export const server = createServer(app);

export default { app, databaseConnection };
