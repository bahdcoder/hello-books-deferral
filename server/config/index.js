import dotenv from "dotenv";
import auth from "@config/auth";
import server from "@config/server";
import database from "@config/database";

dotenv.config();

export default {
  auth: auth(),
  server: server(),
  database: database()
};
