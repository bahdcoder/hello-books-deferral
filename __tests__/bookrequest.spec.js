import faker from 'faker';
import supertest from 'supertest';
import User from '@models/User';
import jwt from 'jsonwebtoken';
import config from '@config';
import { app, databaseConnection } from '@server/app';
import { getUser, createUser } from '@tests/utils/helpers';

const server = () => supertest(app);
const bookrequestRoute = '/api/v1/book-request';

describe('BOOK REQUEST API ENDPOINTS', () => {
    beforeAll(async () => {
        await databaseConnection.migrate.latest();
        await databaseConnection('users').truncate();
    });

    afterAll(async () => {
        await databaseConnection('users').truncate();
    });

    it('should return an error if desciption field is not provided', async () => {
        const { status, body } = await server()
            .post(bookrequestRoute)
            .send();

        expect(status).toBe(422);
        expect(body).toMatchSnapshot();
    });

    it.skip('Should to make a book request', async () => {
        const user = getUser();
        const createdUser = await createUser(user);

        const token = jwt.sign(
            { id: createdUser.id, email: createdUser.email },
            config.auth.secret
        );

        const description = faker.lorem.sentences();

        const { status, body } = await server()
            .post(bookrequestRoute)
            .set('x-access-token', token)
            .send({
                description
            });

        expect(status).toBe(201);
        expect(body).toMatchSnapshot();
    });

    it('should get all book requests', async () => {
        const user = getUser();
        const createdUser = await createUser(user);

        const token = jwt.sign(
            { id: createdUser.id, email: createdUser.email },
            config.auth.secret
        );
        const { status } = await server()
            .get(bookrequestRoute)
            .set('x-access-token', token);

        expect(status).toBe(200);
    });
});
