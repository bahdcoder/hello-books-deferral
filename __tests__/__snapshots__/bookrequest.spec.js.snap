// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BOOK REQUEST API ENDPOINTS Should to make a book request 1`] = `
Object {
  "data": Object {
    "message": "Your request has been noted",
  },
  "status": "success",
}
`;

exports[`BOOK REQUEST API ENDPOINTS should get all book requests 1`] = `
Object {
  "data": Object {
    "message": "Your request has been noted",
  },
  "status": "success",
}
`;

exports[`BOOK REQUEST API ENDPOINTS should return an error if desciption field is not provided 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "description",
      "message": "description is required",
      "validation": "required",
    },
  ],
  "status": "error",
}
`;
